File: android/app/src/main/java/com/vortex/a3/core/notif/LiveActivity.kt
package com.vortex.a3.core.notif

import org.json.JSONObject

/**
 * A "live activity" — an ongoing, in-place-updating phone notification (ride
 * ETA, navigation, delivery, timer, download) mirrored to the laptop's top
 * bar as a persistent pill (live-activity style). Unlike a one-shot
 * [NotificationMirror], the same [key] is re-sent as the source updates, and
 * an [ended] message clears the pill.
 *
 * Tiny by design — fits a single BLE notify. Mirrors Rust
 * `core::live_activity::LiveActivity`.
 */
data class LiveActivity(
    /** Stable identity (Android StatusBarNotification.key) — the same across
     *  every update so the laptop updates one pill in place. */
    val key: String,
    /** Friendly app label (e.g. "Yandex Go"), or the package as a fallback. */
    val app: String = "",
    /** Package name — the laptop looks up this app's cached icon for the
     *  live-activity's own tray icon. */
    val appId: String = "",
    /** Short headline (contentTitle), e.g. "Heading your way". */
    val title: String = "",
    /** Detail line (contentText), e.g. "Arriving in 1 min". */
    val text: String = "",
    /** Extra line (sub-text), e.g. "2.3 km · 6 min". */
    val sub: String = "",
    /** Progress 0..100, or -1 when there's no determinate progress
     *  (indeterminate spinner / category-only live activity). */
    val progress: Int = -1,
    /** True = the activity finished / was removed → clear its pill. */
    val ended: Boolean = false,
    /** Media pill only: whether the player is currently playing. Non-null
     *  marks this activity as a now-playing pill — the laptop shows
     *  prev/play-pause/next transport buttons and picks the ⏸/▶ icon
     *  from this. Null for every non-media live activity. */
    val playing: Boolean? = null,
) {
    fun toJsonBytes(): ByteArray {
        val o = JSONObject()
        o.put("key", key)
        if (app.isNotEmpty()) o.put("app", app)
        if (appId.isNotEmpty()) o.put("app_id", appId)
        if (title.isNotEmpty()) o.put("title", title)
        if (text.isNotEmpty()) o.put("text", text)
        if (sub.isNotEmpty()) o.put("sub", sub)
        o.put("progress", progress)
        if (ended) o.put("ended", true)
        playing?.let { o.put("playing", it) }
        return o.toString().toByteArray(Charsets.UTF_8)
    }

    companion object {
        fun fromJsonBytes(bytes: ByteArray): LiveActivity? = try {
            val o = JSONObject(String(bytes, Charsets.UTF_8))
            LiveActivity(
                key = o.optString("key", ""),
                app = o.optString("app", ""),
                appId = o.optString("app_id", ""),
                title = o.optString("title", ""),
                text = o.optString("text", ""),
                sub = o.optString("sub", ""),
                progress = o.optInt("progress", -1),
                ended = o.optBoolean("ended", false),
                playing = if (o.has("playing")) o.optBoolean("playing") else null,
            )
        } catch (_: Throwable) {
            null
        }
    }
}

File: linux/daemon/src/core/live_activity.rs
//! Live-activity payload (phone → laptop).
//!
//! An ongoing, in-place-updating phone notification (ride ETA, navigation,
//! delivery, timer) forwarded over the Noise-sealed BLE link (frame type
//! `ty::LIVE_ACTIVITY`) and shown as a persistent pill on the laptop's top
//! bar (the system-tray label). Unlike a one-shot notification it carries a
//! stable `key` re-sent on every update, plus an `ended` flag to clear it.

use serde::{Deserialize, Serialize};

fn neg_one() -> i32 {
    -1
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LiveActivity {
    /// Stable identity (Android `StatusBarNotification.key`) — the same
    /// across every update so the laptop refreshes one pill in place.
    #[serde(default)]
    pub key: String,
    /// Friendly app label (e.g. "Yandex Go"), package name as a fallback.
    #[serde(default)]
    pub app: String,
    /// Package name — the laptop looks up this app's cached icon for the
    /// live-activity's own tray icon.
    #[serde(default)]
    pub app_id: String,
    /// Short headline (e.g. "Heading your way").
    #[serde(default)]
    pub title: String,
    /// Detail line (e.g. "Arriving in 1 min").
    #[serde(default)]
    pub text: String,
    /// Extra line (sub-text), e.g. "2.3 km · 6 min".
    #[serde(default)]
    pub sub: String,
    /// Progress 0..=100, or `-1` when there's no determinate progress.
    #[serde(default = "neg_one")]
    pub progress: i32,
    /// Epoch-millis a duration counts from (the in-call pill: when the call
    /// connected). `0` = no timer. The GNOME extension ticks it LOCALLY so the
    /// daemon needn't republish every second (which would starve the D-Bus
    /// interface's method dispatch). Phone-sent pills omit it.
    #[serde(default)]
    pub started_at: i64,
    /// In-call pill audio toggles (the call pill only): `muted` → Mute/Unmute,
    /// `speaker` → Speaker on/off, `has_earbuds` → hide the Speaker button.
    #[serde(default)]
    pub muted: bool,
    #[serde(default)]
    pub speaker: bool,
    #[serde(default)]
    pub has_earbuds: bool,
    /// True = the activity finished / was removed → clear its pill.
    #[serde(default)]
    pub ended: bool,
    /// Now-playing pill only (mirrors Kotlin `LiveActivity.playing`):
    /// `Some(_)` marks this as the phone's media pill — the extension draws
    /// prev/play-pause/next transport buttons and keys ⏸/▶ off the value.
    /// `None` for every non-media live activity.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub playing: Option<bool>,
}
